### DISCUSSION TITLE
How can I setup lsp for java and scala

### DISCUSSION THREAD

**Discussion by Kim-Robin** [LazyVim/LazyVim#347] [Q&A] [answered]
I am very new to mason and lazyvim. (just new to neovim tbh)

I have installed jdtls for java but nothing  happened to my java project. no auto completion and other lsp stuff.
Also, I was not able to find scala lsp on mason.

Can anyone share how to setup java and scala on lazyvim?

**Comment by CinderellaSharrieffG** [comment]
Sure, I can help you with that. Here's how to set up Java and Scala in LazyVim using jdtls and metals, respectively:

Install jdtls:
Open a terminal window and run the following command to install jdtls using mason:
mason install jdtls

Configure jdtls:
Add the following lines to your init.vim file to configure jdtls:
let g:lsp_config = {
      \ 'java': {
      \   'cmd': ['jdtls'],
      \   'filetypes': ['java'],
      \   'root_markers': ['pom.xml', 'build.gradle', '.git'],
      \   'settings': {
      \     'java': {
      \       'format': {
      \         'enabled': v:true
      \       },
      \       'signatureHelp': {
      \         'enabled': v:true
      \       },
      \       'contentProvider': {
      \         'preferred': 'fernflower'
      \       }
      \     }
      \   }
      \ }
      \}

" Set up key bindings
nnoremap <leader>ln :LspInfo<CR>
nnoremap <leader>ll :LspRestart<CR>
nnoremap <leader>lr :LspStop<CR>:LspStart<CR>

Install metals:
Open a terminal window and run the following command to install metals using coursier:
cs install metals

Configure metals:
Add the following lines to your init.vim file to configure metals:
let g:lsp_config = get(g:, 'lsp_config', {})
let g:metals_config = {
      \ 'metals': {
      \   'cmd': ['metals'],
      \   'filetypes': ['scala'],
      \   'root_markers': ['.git'],
      \   'log_level': 'warn',
      \   'settings': {
      \     'metals': {
      \       'icon': '•',
      \       'lineEndings': 'Unix',
      \       'downloadSources': false,
      \       'server': {
      \         'memory': {
      \           'min': '1G',
      \           'max': '2G'
      \         }
      \       },
      \       'showImplicitArguments': true,
      \       'showInferredType': true,
      \       'showImplicitConversionsAndClasses': true,
      \       'completion': {
      \         'autoImport': true,
      \         'snippetAutoIndent': true,
      \         'snippetFallbackStrategy': 'Expand',
      \         'snippetAutoTriggerCharacters': [' ', '.', '(', '['],
      \         'showInferredType': true,
      \         'hint': {
      \           'enable': true
      \         },
      \         'resolve': {
      \           'memberNames': true
      \         }
      \       }
      \     }
      \   }
      \ }
      \}

" Set up key bindings
nnoremap <leader>ln :LspInfo<CR>
nnoremap <leader>ll :LspRestart<CR>
nnoremap <leader>lr :LspStop<CR>:LspStart<CR>

After adding these configurations to your init.vim file, save and exit the file. Then restart neovim and open a Java or Scala project to see the LSP features in action.

**Reply by Kim-Robin** [reply]
Thank you for helping me out.

I have received saying "Conflicting configs: /.config/nvim/init.lua .config/nvim/init.vim". 
Is it possible to implement this in lua. I am using lazyvim starter

I was trying to use this kind of syntax https://www.lazyvim.org/plugins/lsp

**Reply by amaanq** [reply]
Good question, I added Scala and Zig recently to my language extras in my config here: https://github.com/amaanq/nvim-config/blob/master/lua/plugins/extras/lang/scala.lua

You can also use my Java one in the same folder, but I'll admit I practically took that from here: https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/extras/lang/java.lua

You don't have to directly copy what I have, but it should give you an idea on how to extend your config by requiring your own plugins

**Reply by Kim-Robin** [reply]
Thank you for helping me out. I have looked into your scala configuration and I was not able to work with it. Seems like metals is not supported by mason and not downloading it.

I did not have chance the test the java one but it looks like it is going to work.

**Reply by amaanq** [reply]
It appears metals is not (perhaps submit a PR to add it ;) ), you can install it yourself and it should work then

**Reply by Kim-Robin** [reply]
I have metal installed on my system and seems like there is Open issue here https://github.com/williamboman/mason.nvim/issues/369.
I think they are working on it. I am just starting to learn scala and I am like junior dev so I am not sure what to work with this metal lsp and submit PR.

**Reply by amaanq** [reply]
Ahh, seems like the maintainer has stated why metals won't be added, and to use the nvim-metals plugin instead

**Reply by Kim-Robin** [reply]
yes. Things are not as easy as what I have thought. Sadly I got used to vim while I am testing out and not using mouse too much. Now I can't go back to other ide(even with the vim plugins). I want to have debugging and jumping to definition of external library working on vim but I have long way to learn lol.

**Reply by amaanq** [reply]
does nvim-metals not provide these things?

**Reply by Kim-Robin** [reply]
I have not used nvim-metals and doing research on how I can implement on lazyvim. Not gonna lie I am very confused with how to set things up with lazyvim and lua.

**Reply by amaanq** [reply, marked as answer]
If you are using the starter, make a file at plugins/extras/lang/scala.lua and try this:

* note that the metals config setup stuff I copied from their example configuration, tweak as you like!

```lua
return {

  -- Add Scala to treesitter
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      if type(opts.ensure_installed) == "table" then
        vim.list_extend(opts.ensure_installed, { "scala" })
      end
    end,
  },

  {
    "scalameta/nvim-metals",
    dependencies = {
      "nvim-lua/plenary.nvim",
      "mfussenegger/nvim-dap",
    },
    config = function()
      -- Autocmd that will actually be in charging of starting the whole thing
      local nvim_metals_group = vim.api.nvim_create_augroup("nvim-metals", { clear = true })

      local metals_config = require("metals").bare_config()

      -- Example of settings
      metals_config.settings = {
        showImplicitArguments = true,
        excludedPackages = { "akka.actor.typed.javadsl", "com.github.swagger.akka.javadsl" },
      }

      -- *READ THIS*
      -- I *highly* recommend setting statusBarProvider to true, however if you do,
      -- you *have* to have a setting to display this in your statusline or else
      -- you'll not see any messages from metals. There is more info in the help
      -- docs about this
      -- metals_config.init_options.statusBarProvider = "on"

      -- Example if you are using cmp how to make sure the correct capabilities for snippets are set
      metals_config.capabilities = require("cmp_nvim_lsp").default_capabilities()

      metals_config.on_attach = function(_client, _bufnr)
        require("metals").setup_dap()
      end

      vim.api.nvim_create_autocmd("FileType", {
        -- NOTE: You may or may not want java included here. You will need it if you
        -- want basic Java support but it may also conflict if you are using
        -- something like nvim-jdtls which also works on a java filetype autocmd.
        pattern = { "scala", "sbt", "java" },
        callback = function()
          require("metals").initialize_or_attach(metals_config)
        end,
        group = nvim_metals_group,
      })
    end,
  },

  -- Correctly setup lspconfig for Scala 🚀
  {
    "neovim/nvim-lspconfig",
    dependencies = {
      "scalameta/nvim-metals",
    },
    opts = {
      setup = {
        metals = {},
      },
    },
  },
}
```

then in config/lazy.lua, import it with 
```lua
      { import = "plugins.extras.lang.scala" },
```

Give that a shot, it should provide the lsp and debugger stuff

**Reply by Kim-Robin** [reply]
Thank you I will take a look and test it out.👍

**Reply by Kim-Robin** [reply]
I get this weird error using jstl. you have any thoughts on it @amaanq  
I am using https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/extras/lang/java.lua as it is btw

![Screenshot 2023-03-25 011157](https://user-images.githubusercontent.com/41379739/227698163-5fc4f829-833d-4ca9-ab65-c53dc47698be.png)

**Reply by amaanq** [reply]
jdtls requires java 17 and above, are you on that?

**Reply by Kim-Robin** [reply]
Yes I have java 17
